### PR TITLE
Refactor std::tx

### PR DESCRIPTION
**Pull request by simonr0204**
adds functionality to `std::tx` to get script bytecode

Refactored module:

- Std-lib users can get input/output at a given _index_ (they shouldn't need to manually handle pointers)
- break out `get_b256_from_pointer_offset` which is useful everywhere, since most types are either 1 word (so `lw` is sufficient) or 32 bytes. Other functions (e.g. `tx_input_coin_owner`) can re-use this.

### CODE REVIEW

File: sway-lib-std/src/tx.sw [hunk 1]
```diff
@@ -166,6 +166,15 @@ pub fn get_script_data<T>() -> T {
     }
 }
 
+/// Get the script bytecode
+fn get_script_bytecode<T>() -> T {
```
Comment (nfurfaro) [inline sway-lib-std/src/tx.sw:170]: Is there no need to pass in the script bytecode length in addition to the start?

File: sway-lib-std/src/tx.sw [hunk 2]
```diff
@@ -186,12 +195,18 @@ pub fn tx_input_type(ptr: u32) -> u8 {
...
 }
 
-/// If the input's type is `InputCoin`, return the owner.
-/// Otherwise, undefined behavior.
-pub fn tx_input_coin_owner(input_ptr: u32) -> Address {
-    let owner_addr = ~Address::from(asm(buffer, ptr: input_ptr) {
-        // Need to skip over six words, so add 8*6=48
-        addi ptr ptr i48;
+/// Get the type of an input at a given index
+fn get_input_type(index: u8) -> u8 {
```
Comment (nfurfaro) [inline sway-lib-std/src/tx.sw:199]: I like this, already adopting it in the token gateway contract.

File: sway-lib-std/src/tx.sw [hunk 3]
```diff
@@ -186,12 +195,18 @@ pub fn tx_input_type(ptr: u32) -> u8 {
...
+    let ptr = tx_input_pointer(index);
+    let input_type = tx_input_type(ptr);
+    input_type
+}
+
+/// Read 256 bits from memory at a given offset from a given pointer
+fn get_b256_from_pointer_offset(pointer: u32, offset: u32) -> b256 {
```
Comment (nfurfaro) [inline sway-lib-std/src/tx.sw:206]: In order to use this function to replace the current `tx_input_coin_owner()`/`tx_input_message_owner()`, one would need to know:
- the input pointer (already a requirement with the current model), &
- the offset. 
So, we would need to make the offsets available as consts in the stdlib or something.
The current approach I've taken is to create a `tx_input_owner()` function which returns a result, using the `Err` variant for inputs other than `coin` or `message`.
I'm not sure if these offsets will be changing with the introduction of the `gtf` instruction as well?
cc @adlerjohn

Comment (nfurfaro) [reply sway-lib-std/src/tx.sw:206]: Ah, I should have scrolled down more; I see that you've simply wrapped this function with `get_input_coin_owner()` avoiding the need to have users manually pass in offsets.

Comment (nfurfaro) [reply sway-lib-std/src/tx.sw:206]: In general, getting a `b256` value is a common enough operation to warrant this function IMO.

File: sway-lib-std/src/tx.sw (hunk 1 above)
Comment (simonr0204) [reply sway-lib-std/src/tx.sw:170]: I've been handling the length implicitly by casting to a `u64` array of the correct length, so what you actually get is the script data padded to the next whole word. I will document this in the comments.

File: sway-lib-std/src/tx.sw [hunk 4]
```diff
@@ -225,6 +245,13 @@ pub fn tx_output_type(ptr: u32) -> u8 {
     }
 }
 
+/// Get the type of an output at a given index
+fn get_output_type(index: u8) -> u8 {
```
[the remark this answers is not in the document]
Comment (nfurfaro) [reply sway-lib-std/src/tx.sw:249]: same for `get_input_type()`

[the remark this answers is not in the document]
Comment (simonr0204) [reply sway-lib-std/src/tx.sw]: Resolved by [357182b](https://github.com/FuelLabs/sway/pull/2142/commits/357182b7379b764ea3b79d65809bd54397d60e3d)

File: sway-lib-std/src/tx.sw (hunk 1 above)
Comment (simonr0204) [reply sway-lib-std/src/tx.sw]: Resolved by [357182b](https://github.com/FuelLabs/sway/pull/2142/commits/357182b7379b764ea3b79d65809bd54397d60e3d)

File: sway-lib-std/src/tx.sw [hunk 5]
```diff
@@ -166,6 +166,17 @@ pub fn get_script_data<T>() -> T {
...
 
+/// Get the script bytecode
+/// Must be cast to a u64 array, with sufficient length to contain the bytecode.
+/// Bytecode will be padded to next whole word.
+pub fn tx_script_bytecode<T>() -> T {
+    let script_ptr = tx_script_start_offset();
+    let script_bytecode = asm(r1: script_ptr) {
+        r1: T
+    };
+    script_bytecode
```
Comment (otrho) [inline sway-lib-std/src/tx.sw:177]:
```suggestion
   asm(r1: script_ptr) {
        r1: T
    }
```

File: sway-lib-std/src/tx.sw [hunk 6]
```diff
@@ -179,19 +190,25 @@ pub fn tx_input_pointer(index: u64) -> u32 {
...
-/// Otherwise, undefined behavior.
-pub fn tx_input_coin_owner(input_ptr: u32) -> Address {
-    let owner_addr = ~Address::from(asm(buffer, ptr: input_ptr) {
-        // Need to skip over six words, so add 8*6=48
-        addi ptr ptr i48;
+/// Get the type of an input at a given index
+pub fn tx_input_type(index: u64) -> u8 {
+    let ptr = tx_input_pointer(index);
+    let input_type = tx_input_type_from_pointer(ptr);
+    input_type
```
Comment (otrho) [inline sway-lib-std/src/tx.sw:204]:
```suggestion
    tx_input_type_from_pointer(ptr)
```

File: sway-lib-std/src/tx.sw [hunk 7]
```diff
@@ -200,8 +217,14 @@ pub fn tx_input_coin_owner(input_ptr: u32) -> Address {
...
-
+    }
+}
+/// If the input's type is `InputCoin`, return the owner.
+/// Otherwise, undefined behavior.
+pub fn tx_input_coin_owner(index: u64) -> Address {
+    let input_ptr = tx_input_pointer(index);
+    // Need to skip over six words, so offset is 8*6=48
+    let owner_addr = ~Address::from(b256_from_pointer_offset(input_ptr, 48));
     owner_addr
```
Comment (otrho) [inline sway-lib-std/src/tx.sw:228]:
```suggestion
    ~Address::from(b256_from_pointer_offset(input_ptr, 48))
```

File: sway-lib-std/src/tx.sw [hunk 8]
```diff
@@ -218,17 +241,25 @@ pub fn tx_output_pointer(index: u64) -> u32 {
...
         lw r1 r2 i0;
         r1: u8
     }
 }
 
+/// Get the type of an output at a given index
+pub fn tx_output_type(index: u64) -> u8 {
+    let ptr = tx_output_pointer(index);
+    let output_type = tx_output_type_from_pointer(ptr);
+    output_type
```
Comment (otrho) [inline sway-lib-std/src/tx.sw:255]:
```suggestion
    tx_output_type_from_pointer(ptr)
```

File: test/src/sdk-harness/test_projects/script_bytecode/Forc.toml [hunk 9]
```diff
@@ -0,0 +1,8 @@
+[project]
+authors = ["Simon"]
```
Comment (otrho) [inline test/src/sdk-harness/test_projects/script_bytecode/Forc.toml:2]:
```suggestion
authors = ["Fuel Labs <contact@fuel.sh>"]
```

**Review by otrho — COMMENTED** [review]
Mostly aesthetics I guess -- places where you assign a value and then just return it.  Might as well just return the value.

File: sway-lib-std/src/tx.sw (hunk 5 above)
Comment (simonr0204) [reply sway-lib-std/src/tx.sw]: Resolved in [a7c7a91](https://github.com/FuelLabs/sway/pull/2142/commits/a7c7a91776162bd5d4a0bbbbf94c14c066c83e9a)

File: sway-lib-std/src/tx.sw (hunk 8 above)
Comment (simonr0204) [reply sway-lib-std/src/tx.sw]: Resolved in [a7c7a91](https://github.com/FuelLabs/sway/pull/2142/commits/a7c7a91776162bd5d4a0bbbbf94c14c066c83e9a)

File: sway-lib-std/src/tx.sw (hunk 7 above)
Comment (simonr0204) [reply sway-lib-std/src/tx.sw]: Resolved in [a7c7a91](https://github.com/FuelLabs/sway/pull/2142/commits/a7c7a91776162bd5d4a0bbbbf94c14c066c83e9a)

**Review by otrho — APPROVED** [review]

File: test/src/sdk-harness/test_projects/script_bytecode/mod.rs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    let script_hash = Hasher::hash(&script_bytecode); // This is the hard that must be hard-coded in the predicate
+
+    // Run script and get the hash it returns
+    let path_to_bin = "test_projects/script_bytecode/out/debug/script_bytecode.bin";
+    let return_val = run_compiled_script(path_to_bin).await.unwrap();
+    let script_hash_from_vm = unsafe{Bytes32::from_slice_unchecked(return_val[0].data().unwrap())};
+
+    assert_eq!(script_hash_from_vm, script_hash);
+
+}
```
Comment (adlerjohn) [inline test/src/sdk-harness/test_projects/script_bytecode/mod.rs:50]: fmt

Comment (simonr0204) [reply test/src/sdk-harness/test_projects/script_bytecode/mod.rs]: Resolved [7784f85](https://github.com/FuelLabs/sway/pull/2142/commits/7784f8557b472e31cfbc9bc6d657b5266a721495)

**Review by otrho — APPROVED** [review]

**Review by adlerjohn — APPROVED** [review]

**Pull request merged**